src/transport/SessionManager.cpp

```diff
@@ -773,29 +786,59 @@ void SessionManager::SecureGroupMessageDispatch(const PacketHeader & packetHeade
         return;
     }
 
-    System::PacketBufferHandle msgCopy;
+    // Extract MIC from the end of the message.
+    uint8_t * data     = msg->Start();
+    uint16_t len       = msg->DataLength();
+    uint16_t footerLen = packetHeader.MICTagLength();
+    VerifyOrReturn(footerLen <= len);
+
+    uint16_t taglen = 0;
+    MessageAuthenticationCode mac;
+    ReturnOnFailure(mac.Decode(packetHeader, &data[len - footerLen], footerLen, &taglen));
+    VerifyOrReturn(taglen == footerLen);
+
     bool decrypted = false;
     while (!decrypted && iter->Next(groupContext))
     {
+        CryptoContext context(groupContext.keyContext);
+        msgCopy = msg.CloneData();
+
+        if (packetHeader.HasPrivacyFlag())
+        {
+            // Perform privacy deobfuscation, if applicable.
+            uint8_t * privacyHeader = msgCopy->Start() + Header::kPrivacyHeaderOffset;
```

Comment:
Can we make this a (static?) method on PacketHeader or something?  It's weird to be getting the length from the PacketHeader but the data from somewhere else.

Maye DecodeFixed should stash this data and then we can just ask the PacketHeader instance for it?

Reply:
`DecodeFixed` is actually storing parsed header fields that can be reused for all copies of that message.

The pointer to the start of the privacyHeader is unique to each msgCopy and needs to be dynamically calculated in order to deobfuscate so we can call `DecodeAndConsume`.

Reply:
Ah, indeed.  This still feels like it would be better as a static PacketHeader method....

Reply:
Added a fix.